src/Core/Monitor/IApplicationUserResolver.cs

```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Security.Principal;
+using System.Text;
+
+namespace Gibraltar.Monitor
+{
+    /// <summary>
+    /// Implemented to translate an IPrincipal to a Loupe ApplicationUser
+    /// </summary>
+    public interface IApplicationUserResolver
```

Comment:
IMO this isn’t really a “resolver” since you’re providing an `IPrincipal` already obtained out-of-band as input to the `ResolveApplicationUser()` call. I’d consider renaming to something like `IApplicationPrincipalAdapter` or maybe `...Provider`.

I also couldn’t find an implementation - more out of curiosity than anything else, where is it?

Reply:
Provider sounds good - because the idea is that you inject your code to take an IPrinciple and record a much more detailed object from it (an "ApplicationUser").  

I didn't create a generic ASP.NET Core implementation yet; each app would have to make its own based on the claims and structure of its own user storage (which is what we required previously as well, just through a different API) but possibly we could do better if there's a standard "shape" to most ASP.NET Core authentication.  You can see two example implementations in the unit tests for ApplicationUserTests, e.g.:

```cs
        private class ResolveUserForCurrentPrincipal : IApplicationUserResolver
        {
            public ApplicationUser ResolveApplicationUser(IPrincipal principal, Func<ApplicationUser> userFactory)
            {
                var identity = principal.Identity;
                var newUser = userFactory();
                newUser.Caption = Environment.GetEnvironmentVariable("USERNAME");
                newUser.Organization = "Unit test";
                newUser.EmailAddress = "support@gibraltarsoftware.com";
                newUser.Phone = "443 738-0680";
                newUser.Properties.Add("Customer Key", "1234-5678-90");
                newUser.Properties.Add("License Check", null);

                return newUser;
            }
        }
```
This implementation is clearly bogus and for a unit test; you wouldn't pull the user name from the environment, you wouldn't have everything be hard coded, etc.   

What do you think of the user factory approach here?  I'm not entirely thrilled with it but it does hit the goal of providing a call to get a "blank" ApplicationUser (that under the covers has the right key associated with it and is safe for the system to serialize).

Reply:
Interface renamed, changed method to Try* signature to match the IPrincipalProvider change and then switched to a Lazy<T> to eliminate the factory. Let me know what you think!

Reply:
Reads much better to me with these names and also really like the `Try...` suggestion from Mark. Looking good.